model/transformer.py

```diff
@@ -0,0 +1,124 @@
+  """Transformer Architecture using Vanilla Transformer.
+  A sequence of tokens are passed to the Embedding layer.
+  After, Embeddings are passed through Positional Encoding layer.
+  The logits need to be passed through final Linear layer.
+  """
+  def __init__(self,
+               input_vocab_size: int,
+               output_vocab_size: int,
+               # config CONCEPT_IDENTIFICATION.LSTM_BASED
+               config: CfgNode,
+               dropout=0.5,
+               device="cpu"):
+    super(TransformerSeq2Seq, self).__init__()
+    hidden_size = config.HIDDEN_SIZE
+    embedding_dim = config.EMB_DIM
+    head_number = config.NUM_HEADS
+    # Embed inputs
+    self.enc_embedding = nn.Embedding(input_vocab_size, embedding_dim).to(device)
+    self.dec_embedding = nn.Embedding(output_vocab_size, embedding_dim).to(device)
+    # Positional Encoding
+    self.pos_encoder = PositionalEncoding(embedding_dim, dropout)
+    self.pos_decoder = PositionalEncoding(embedding_dim, dropout)
+    # Vanilla Transformer
+    self.transformer = Transformer(embedding_dim, head_number, hidden_size,
+                                   dropout=dropout)
+    # Linear Transformation of Transformer Output
+    self.dense = nn.Linear(embedding_dim, output_vocab_size).to(device)
+    # Device to train on
+    self.device = device
+    # Masks
+    self.trg_mask = None
+
+
+  def make_triangular_mask(self, inp):
+    """
+    Wrapper for transformer mask (upper triangular)
+    """
+    return  nn.Transformer().generate_square_subsequent_mask(inp.size(0)).to(self.device)
+
+  def forward(self,
+              input_sequence: torch.Tensor,
+              input_lengths: torch.Tensor = None,
+              gold_output_sequence: torch.Tensor = None,
+              max_out_length: int = None):
+    """Forward transformer
+
+    Args:
+        input_sequence (torch.Tensor): encoder inputs
+        input_lengths (torch.Tensor): inputs length
+        gold_output_sequence (torch.Tensor): decoder inputs
+        max_out_length (int): maximum output length
+    Returns:
+        logits of shape ...
+        predictions of shape (out seq len, batch size)
+    """
+    input_seq_len = input_sequence.shape[0]
+    input_seq_size = input_sequence.shape[1]
+
+    input_sequence = self.enc_embedding(input_sequence)
+    input_sequence = self.pos_encoder(input_sequence)
+    if self.training:
+      gold_output_sequence = self.dec_embedding(gold_output_sequence)
+      gold_output_sequence = self.pos_decoder(gold_output_sequence)
+      self.trg_mask = self.make_triangular_mask(gold_output_sequence)
+      transformer_out = self.transformer(input_sequence,
+                                         gold_output_sequence,
+                                         tgt_mask=self.trg_mask)
+      logits = self.dense(transformer_out)
+      activated_outputs = torch.softmax(logits, dim=-1)
+      predictions = torch.argmax(activated_outputs, dim=-1)
```

Comment:
this code could be extracted in a component; it's almost completely duplicated on the else branch (just send in the mask from the outside maybe)

Reply:
extracted in _apply_layer_ function